### PR TITLE
Implement EncryptionKeyRotation spec

### PR DESCRIPTION
**Pull request by black-dragon74**
This patch implements the EncryptionKeyRotation spec for ceph-csi

### CODE REVIEW

File: internal/util/crypto.go
Comment (black-dragon74) [inline internal/util/crypto.go]: TODO: `returns` instead of `return`

File: internal/util/file.go
Comment (nixpanic) [inline internal/util/file.go]: maybe move `createTempFile` from `internal/kms/vault.go` to a new `internal/util/file` package?

File: internal/rbd/rbd_util.go
Comment (nixpanic) [inline internal/rbd/rbd_util.go]: move this to `internal/rbd/encryption.go`?

File: internal/rbd/rbd_util.go
Comment (nixpanic) [inline internal/rbd/rbd_util.go]: why not?

Will you add support for `fscrypt` later on?

**Review by nixpanic — COMMENTED** [review]
Just added a few small comments. Also try to create unit-tests for the helper functions you introduce.

File: internal/rbd/rbd_util.go
Comment (black-dragon74) [reply internal/rbd/rbd_util.go]: Yes.

File: internal/rbd/rbd_util.go
Comment (black-dragon74) [reply internal/rbd/rbd_util.go]: Ack.

File: internal/util/file.go
Comment (black-dragon74) [reply internal/util/file.go]: Umm. It's already there, no?

Comment (nixpanic) [reply internal/util/file.go]: there is already a `createTempFile()` function in `internal/kms/vault.go`. Move that around and rename it to `CreateTempFile` so that there is less duplication.

Comment (black-dragon74) [reply internal/util/file.go]: Ah, I see. I thought you were referring to the current one 😂. Will take care of it.

**Comment by black-dragon74** [comment]
Note: The CI failures are expected until: https://github.com/csi-addons/spec/pull/65 is merged.

File: internal/util/file.go
Comment (black-dragon74) [reply internal/util/file.go]: Well, turns out, importing utils in kms will lead to circular dependency. Thoughts?

Comment (nixpanic) [reply internal/util/file.go]: yes, move the function to `internal/util/file/temp.go` and have a new `file` package.

**Comment by black-dragon74** [comment]
I’ll take care of the failures since the linked PR is merged :)

**Comment by iPraveenParihar** [comment]
@black-dragon74, please follow these [conventions](https://github.com/ceph/ceph-csi/blob/devel/docs/development-guide.md#commit-messages) for the commit messages.

File: internal/kms/vault_test.go
Comment (nixpanic) [inline internal/kms/vault_test.go]: move this test to the `github.com/ceph/ceph-csi/internal/util/file` package. Or remove it in case this is already tested there.

**Comment by black-dragon74** [comment]
> @black-dragon74, please follow these [conventions](https://github.com/ceph/ceph-csi/blob/devel/docs/development-guide.md#commit-messages) for the commit messages.

Thanks Praveen. I'll be squashing all the commits into one before merging.

**Review by Rakshith-R — COMMENTED** [review]
- Add shared volumelocks with nodeserver and acquire it before proceeding with rotation ?
- Block rotation for RWX volumes ?
- Add testing tool at csi addons and paste manual testing logs or add e2e here to test the procedure.

**Comment by black-dragon74** [comment]
> * Add shared volumelocks with nodeserver and acquire it before proceeding with rotation ?

We are sending the request to only one node. Is volume lock required in this case?

> * Block rotation for RWX volumes ?

Totally missed it, will do :)

> * Add testing tool at csi addons and paste manual testing logs or add e2e here to test the procedure.

WIP

**Comment by Rakshith-R** [comment]
> > * Add shared volumelocks with nodeserver and acquire it before proceeding with rotation ?
> 
> We are sending the request to only one node. Is volume lock required in this case?

To prevent nodeunstage when rotating keys. 

> 
> > * Block rotation for RWX volumes ?
> 
> Totally missed it, will do :)
> 
> > * Add testing tool at csi addons and paste manual testing logs or add e2e here to test the procedure.
> 
> WIP

File: internal/csi-addons/rbd/encryptionkeyrotation.go
Comment (Rakshith-R) [inline internal/csi-addons/rbd/encryptionkeyrotation.go]: this should be at line 54 ?

File: internal/rbd/encryption.go
Comment (Rakshith-R) [inline internal/rbd/encryption.go]: why ?

Comment (black-dragon74) [reply internal/rbd/encryption.go]: The key has been removed from the KMS and doesn't exist anywhere but the backup slot. Should we error out here?

I returned nil coz it will be overwritten in the next run anyways and there is no way to fetch the old key anymore.

EDIT:  Let's return error here. Coz leaving the key in there somehow defeats the purpose of rotating the key in the 1st place.

File: internal/csi-addons/rbd/encryptionkeyrotation.go
Comment (black-dragon74) [reply internal/csi-addons/rbd/encryptionkeyrotation.go]: Umm.. We are not doing any ops on the volume that requires the use of locks until line 74. I will happily move it if I am missing something here. Please let me know :)

Comment (Rakshith-R) [reply internal/csi-addons/rbd/encryptionkeyrotation.go]: There's no need for those ops at all if another process is holding the locks over the volume.
You need to use the share the same lock object used by nodeserver to prevent nodeunstage calls during key rotation ops.

Comment (black-dragon74) [reply internal/csi-addons/rbd/encryptionkeyrotation.go]: Makes sense, totally. Thanks Rakshith!

File: internal/csi-addons/rbd/encryptionkeyrotation.go
Comment (nixpanic) [inline internal/csi-addons/rbd/encryptionkeyrotation.go]: you don't need anything from the NodeServer, except the `ns.VolumeLocks`. It would be cleaner to pass the `*util.VolumeLocks` instead of the whole NodeServer.

File: internal/util/cryptsetup.go
Comment (nixpanic) [inline internal/util/cryptsetup.go:196]: consider logging the stderr output as well, maybe not included in the error, but as a 'normal' log

Comment (black-dragon74) [reply internal/util/cryptsetup.go:196]: Since this is the only instance where we need to log, is it worth it to pass the logger instance to this function as rest of the code in this file is decoupled. If we need to log it anyways, let's include it in the returned error itself?

Comment (nixpanic) [reply internal/util/cryptsetup.go:196]: It might help to have the log in case something went wrong. I do not know if the `err` includes (some parts of) `stderr` already. Including the whole `stderr` in `err` might be a bit much, it could be multiple lines.

You can pass the `context.Context` from the caller, and use `internal/util/log.ErrorLog()` or similar, or `ErrorLogMsg()` if you do not want/can pass the context around.

**Comment by nixpanic** [comment]
How does this relate to #4655 ?

**Comment by black-dragon74** [comment]
> How does this relate to #4655 ?

The linked PR is documentation for this PR.

File: internal/util/cryptsetup.go
Comment (black-dragon74) [reply internal/util/cryptsetup.go:196]: > I do not know if the `err` includes (some parts of) `stderr` already.

It does not. I will add a log statement here :)

File: internal/csi-addons/rbd/encryptionkeyrotation.go
Comment (Madhu-1) [inline internal/csi-addons/rbd/encryptionkeyrotation.go]: why this is marked as Aborted for all the err cases?

File: internal/rbd/encryption.go
Comment (Madhu-1) [inline internal/rbd/encryption.go]: this error message is not clear, can you please update this one?

File: internal/rbd/encryption.go
Comment (Madhu-1) [inline internal/rbd/encryption.go:457]: can you please update this to block RWX rbd volume as well.

File: internal/util/cryptsetup.go
Comment (Madhu-1) [inline internal/util/cryptsetup.go]: i think it's not required, stdout will be set only during success state

File: internal/util/cryptsetup.go
Comment (Madhu-1) [inline internal/util/cryptsetup.go]: log err as well

File: internal/util/file/file.go
Comment (Madhu-1) [inline internal/util/file/file.go]: do we need this FIXME here any discussion is required?

File: internal/util/file/file.go
Comment (Madhu-1) [inline internal/util/file/file.go]: declare `c` above and change `:=` to `=`

File: internal/util/file/file_test.go
Comment (Madhu-1) [inline internal/util/file/file_test.go]: move this to defer after creating file

Comment (Madhu-1) [reply internal/util/file/file_test.go]: take care of this in all other places

File: internal/csi-addons/rbd/encryptionkeyrotation.go
Comment (Madhu-1) [inline internal/csi-addons/rbd/encryptionkeyrotation.go]: i think this order need to be changed, please check https://github.com/ceph/ceph-csi/blob/devel/docs/coding.md#imports

File: internal/csi-addons/rbd/encryptionkeyrotation.go
Comment (Madhu-1) [inline internal/csi-addons/rbd/encryptionkeyrotation.go]: update this comment to get the rbdVolume or remove this one?

File: internal/rbd/encryption.go
Comment (black-dragon74) [reply internal/rbd/encryption.go:457]: Adding the check here would require us to change the spec to add support for volume capabilities. Since we will be supporting RWX soon, I have implemented this check in csi-addons.

File: internal/csi-addons/rbd/encryptionkeyrotation.go
Comment (black-dragon74) [reply internal/csi-addons/rbd/encryptionkeyrotation.go]: Changing to `codes.InvalidArgument`

File: internal/util/file/file.go
Comment (black-dragon74) [reply internal/util/file/file.go]: Removed.

File: internal/csi-addons/rbd/encryptionkeyrotation.go
Comment (Madhu-1) [reply internal/csi-addons/rbd/encryptionkeyrotation.go]: it should be internal?

File: internal/rbd/encryption.go
Comment (Madhu-1) [reply internal/rbd/encryption.go:457]: we cannot add a check on the csi-addons, we can document this as an unsupported feature (this can lead to other problem as we are supporting auto rotation) or change the spec to add this one. 

@nixpanic WDYT?

File: internal/util/file/file.go
Comment (black-dragon74) [reply internal/util/file/file.go]: Done. May I know the benefit of doing it this way?

Comment (Madhu-1) [reply internal/util/file/file.go]: just to avoid variable shadowing

File: internal/rbd/encryption.go
Comment (black-dragon74) [reply internal/rbd/encryption.go:457]: If we decide to add it, a PR is open at: https://github.com/csi-addons/spec/pull/70

Comment (Madhu-1) [reply internal/rbd/encryption.go:457]: @nixpanic ^^

File: internal/rbd/encryption.go
Comment (Madhu-1) [inline internal/rbd/encryption.go:479]: define const for `0` and `1` slots and use the const

**Review by Madhu-1 — COMMENTED** [review]
LGTM, expect pending RWX check

File: internal/csi-addons/rbd/encryptionkeyrotation.go
Comment (iPraveenParihar) [inline internal/csi-addons/rbd/encryptionkeyrotation.go]:
```suggestion
		return nil, status.Error(codes.InvalidArgument, "empty volume ID in request")
```

File: internal/rbd/encryption.go
Comment (iPraveenParihar) [inline internal/rbd/encryption.go]:
```suggestion
// RotateEncryptionKey processes the key rotation for the RBD Volume.
```

File: internal/rbd/encryption.go
Comment (iPraveenParihar) [inline internal/rbd/encryption.go]: better to have consistent error messages across project
```suggestion
		return fmt.Errorf("failed to check encryption state: %w", err)
```

File: internal/rbd/encryption.go
Comment (iPraveenParihar) [inline internal/rbd/encryption.go]:
```suggestion
		return fmt.Errorf("failed to get the device path for %q: %w", rv, err)
```

File: internal/rbd/encryption.go
Comment (iPraveenParihar) [inline internal/rbd/encryption.go]:
```suggestion
		return fmt.Errorf("failed to fetch passphrase for %q: %w",rv, err)
```

similarly take care for below error messages.

**Review by iPraveenParihar — COMMENTED** [review]
nit

File: internal/csi-addons/rbd/encryptionkeyrotation.go
Comment (Madhu-1) [inline internal/csi-addons/rbd/encryptionkeyrotation.go]: this should be internal error

Comment (black-dragon74) [reply internal/csi-addons/rbd/encryptionkeyrotation.go]: Or `InvalidArgument`?

Comment (Madhu-1) [reply internal/csi-addons/rbd/encryptionkeyrotation.go]:
```go
if err != nil {
		switch {
		case errors.Is(err, ErrImageNotFound):
			err = status.Errorf(codes.NotFound, "volume ID %s not found", volID)
		case errors.Is(err, util.ErrPoolNotFound):
			log.ErrorLog(ctx, "failed to get backend volume for %s: %v", volID, err)
			err = status.Errorf(codes.NotFound, err.Error())
		default:
			err = status.Errorf(codes.Internal, err.Error())
		}
```

File: internal/rbd/encryption.go
Comment (Madhu-1) [inline internal/rbd/encryption.go]: add space before %w

File: internal/rbd/encryption.go
Comment (Madhu-1) [inline internal/rbd/encryption.go]: add space before %w

File: internal/rbd/encryption.go
Comment (Madhu-1) [inline internal/rbd/encryption.go]: use luksSlot0

File: internal/rbd/encryption.go
Comment (Madhu-1) [inline internal/rbd/encryption.go]: use luksSlot1

File: internal/rbd/encryption.go
Comment (Madhu-1) [inline internal/rbd/encryption.go]: use luksSlot1

**Review by Madhu-1 — COMMENTED** [review]
nits

**Review by Madhu-1 — APPROVED** [review]

**Review by nixpanic — APPROVED** [review]

**Comment by nixpanic** [comment]
This can be queued for merging once #4655 is merged:

1. use `@mergifyio rebase` to rebase the PR on latest devel branch
2. wait until the Mergify bot has rebased the PR
3. use `@mergifyio queue` to trigger CI jobs and get it merged

**Comment by nixpanic** [comment]
@mergifyio rebase

**Comment by nixpanic** [comment]
@mergifyio queue

**Pull request merged**
